apps/bttester/src/btp_bap.c

```diff
@@ -0,0 +1,454 @@
+
+#define BROADCAST_ADV_INSTANCE                 1
+#define BROADCASTER_INTERRUPT_TASK_PRIO        4
+#define BROADCASTER_INTERRUPT_TASK_STACK_SZ    512
+
+static struct ble_audio_big_subgroup big_subgroup;
+static uint16_t max_sdu;
+static uint32_t sdu_interval;
+
+static uint8_t id_addr_type;
+
+static struct ble_audio_base tester_base;
+
+static os_membuf_t bis_mem[
+    OS_MEMPOOL_SIZE(MYNEWT_VAL(BLE_MAX_BIS),
+                    sizeof(struct ble_audio_bis))
+];
+static struct os_mempool bis_pool;
+
+static os_membuf_t codec_spec_mem[
+    OS_MEMPOOL_SIZE(MYNEWT_VAL(BLE_MAX_BIS) * 2, 19)
+];
+static struct os_mempool codec_spec_pool;
+
+static uint16_t bis_handles[MYNEWT_VAL(BLE_MAX_BIS)];
+/* The timer callout */
+static struct os_callout audio_broadcast_callout;
+
+static int audio_data_offset;
+static struct os_task broadcaster_interrupt_task_str;
+static struct os_eventq broadcaster_interrupt_eventq;
+static os_stack_t broadcaster_interrupt_task_stack[BROADCASTER_INTERRUPT_TASK_STACK_SZ];
+
+static void
+broadcaster_interrupt_task(void *arg)
+{
+    while (1) {
+        os_eventq_run(&broadcaster_interrupt_eventq);
+    }
+}
+
+
+
+static uint8_t
+supported_commands(const void *cmd, uint16_t cmd_len,
+                   void *rsp, uint16_t *rsp_len)
+{
+    struct btp_bap_read_supported_commands_rp *rp = rsp;
+
+    /* octet 0 */
+    tester_set_bit(rp->data, BTP_BAP_READ_SUPPORTED_COMMANDS);
+    tester_set_bit(rp->data, BTP_BAP_BROADCAST_SOURCE_SETUP);
+    tester_set_bit(rp->data, BTP_BAP_BROADCAST_SOURCE_RELEASE);
+    tester_set_bit(rp->data, BTP_BAP_BROADCAST_ADV_START);
+    tester_set_bit(rp->data, BTP_BAP_BROADCAST_ADV_STOP);
+
+    /* octet 1 */
+    tester_set_bit(rp->data, BTP_BAP_BROADCAST_SOURCE_START);
+    tester_set_bit(rp->data, BTP_BAP_BROADCAST_SOURCE_STOP);
+
+    *rsp_len = sizeof(*rp) + 2;
+
+    return BTP_STATUS_SUCCESS;
+}
+
+static int
+broadcast_destroy_fn(struct ble_audio_base *base, void *args)
+{
+    struct ble_audio_bis *bis;
+
+    STAILQ_FOREACH(bis, &big_subgroup.bises, next) {
+        os_memblock_put(&codec_spec_pool, bis->codec_spec_config);
+        os_memblock_put(&bis_pool, bis);
+    }
+
+    memset(&big_subgroup, 0, sizeof(big_subgroup));
+
+    return 0;
+}
+
+static int
+base_create(const struct bap_broadcast_source_setup_cmd *cmd)
+{
+    struct ble_audio_bis *bis;
+    uint16_t sampling_freq = cmd->cc_ltvs[0];
+    uint16_t frame_duration = cmd->cc_ltvs[1];
+    uint16_t chan_loc = BLE_AUDIO_LOCATION_FRONT_LEFT |
+                        BLE_AUDIO_LOCATION_FRONT_RIGHT;
+    uint8_t codec_spec_config[] = 
+        BLE_AUDIO_BUILD_CODEC_CONFIG(sampling_freq, frame_duration, chan_loc,
+                                     max_sdu * 2, );
+
+    tester_base.broadcast_id = 0x42;
+    tester_base.presentation_delay = 20000;
+
+    big_subgroup.bis_cnt = MYNEWT_VAL(BROADCASTER_CHAN_NUM);
+
+    /** LC3 */
+    big_subgroup.codec_id.format = 0x06;
+
+    big_subgroup.codec_spec_config_len = 0;
```

Comment:
Minor comment: we should set `codec_spec_config` to subgroup, and leave `bis` without one. So move it 1 level up

Reply:
Is this necessary to have a whole subgroup set config? Maybe this would be better so we can add specified bises if tests demand it?

Reply:
According to spec, afaik, it's correct to have subgroup config and only bises with different one should have it.

Reply:
BAP spec, 3.7.2.2. Rule 4:

Codec_Specific_Configuration parameters shall be present at Level 2 and may be present at
Level 3. If an identical Codec_Specific_Configuration parameter value is present at Level 2 and at
Level 3, the Codec_Specific_Configuration parameter value at Level 3 shall be treated as the only
instance of that Codec_Specific_Configuration parameter value present. Where a
Codec_Specific_Configuration parameter value includes length-type-value (LTV) structures, an LTV
structure shall be considered an identical parameter to another LTV structure with the same Type
field value, and the Value field of the LTV structure at Level 3 shall be treated as the only instance of
that Value field that is present.